sklearn/decomposition/cdnmf_fast.pyx

```diff
@@ -0,0 +1,117 @@
+# cython: cdivision=True
+# cython: boundscheck=False
+# cython: wraparound=False
+
+# Author: Mathieu Blondel, Tom Dupre la Tour
+# License: BSD 3 clause
+
+import numpy as np
+cimport cython
+cimport numpy as np
+from libc.math cimport fabs, fmax, fmin
+
+
+def _update_cdnmf_fast(double[:, ::1] W, double[:, :] HHt, double[:, :] XHt):
+
+    cdef double violation = 0
+    cdef int n_components = W.shape[1]
+    cdef int n_samples = W.shape[0]  # n_features for H update
+    cdef double grad, pg, hess
+    cdef int i, t, r
+
+    with nogil:
+        for i in range(n_samples):
```

Comment:
Could you add a shuffle option and iterate over rng.permutation(n_samples) here? (i.e., cyclic selection with permutation)